src/modules/artnet/consumer/artnet_consumer.cpp

```diff
@@ -0,0 +1,291 @@
+
+using namespace boost::asio;
+using namespace boost::asio::ip;
+
+namespace caspar {
+    namespace artnet {
+
+        struct configuration
+        {
+            int universe = 0;
+            std::wstring host = L"127.0.0.1";
+            unsigned short port = 6454;
+
+            int refreshRate = 10;
+
+            std::vector<fixture> fixtures;
+        };
+
+        struct artnet_consumer : public core::frame_consumer
+        {
+
+            const configuration config;
+            std::vector<computed_fixture> computed_fixtures;
+
+         public:
+                // frame_consumer
+
+                explicit artnet_consumer(configuration config)
+                    :   config(std::move(config)),
+                        io_service(),
+                        socket(io_service)
+                {
+                    socket.open(udp::v4());
+
+                    std::string host_ = u8(this->config.host);
+                    remote_endpoint = boost::asio::ip::udp::endpoint(boost::asio::ip::address::from_string(host_), this->config.port);
+
+                    compute_fixtures();
+                }
+
+
+                void initialize(const core::video_format_desc& /*format_desc*/, int /*channel_index*/) override {
+                    thread_ = std::thread([this] {
+                        try {
+                            long long time = 1000 / config.refreshRate;
+                            auto last_send = std::chrono::system_clock::now();
+
+                            while (!abort_request_) {
+                                auto now = std::chrono::system_clock::now();
+                                std::chrono::duration<double> elapsed_seconds = now - last_send;
+                                long long elapsed_ms = std::chrono::duration_cast<std::chrono::milliseconds>(elapsed_seconds).count();
+
+                                long long sleep_time = time - elapsed_ms * 1000;
+                                if (sleep_time > 0) std::this_thread::sleep_for(std::chrono::milliseconds(sleep_time));
+
+                                last_send = now;
+
+                                frame_mutex_.lock();
+                                auto frame = last_frame_;
+
+                                frame_mutex_.unlock();
+                                if (!frame) continue; // No frame available
+
+                                uint8_t dmx_data[512];
+                                memset(dmx_data, 0, 512);
+
+                                for (auto computed_fixture : computed_fixtures) {
+                                    auto color = average_color(frame, computed_fixture.rectangle);
+                                    uint8_t* ptr = dmx_data + computed_fixture.address;
+
+                                    switch (computed_fixture.type) {
+                                        case FixtureType::DIMMER:
+                                            ptr[0] = (color.r + color.g + color.b) / 3;
```

Comment:
I believe this implementation to be slightly naive. It would be better to convert to luminance.
```suggestion
                                            ptr[0] =  0.279 * color.r + 0.547 * color.g + 0.106 * color.b
```